### ISSUE TITLE
Streamlink GUI processes fail to close

### ISSUE THREAD

**Issue by adrianAnyansi** [streamlink/streamlink-twitch-gui#396]
Streamlink GUI fails to close properly, leaving the processes frozen.

Reproduction: Time-related? Still investigating

Environment details: Windows 7 Home Premium SP1, Streamlink GUI v1.1.1, Using VLC as stream viewer

### Comments, logs, screenshots, etc.
When this occurs I cannot re-open the GUI as I suppose Windows assumes the GUI is already running. Might be based on time, as open and closing immediately will work properly, but closing after a while fails. Currently the way I solve this is by using End Process via Task Manager.

Tested this with the option GUI->OS integration to make sure the app wasn't minimizing to the tray. If Both is enabled, the Tray Icon remains but is unresponsive.

![screenshot 2017-01-18 09 39 55 - copy](https://cloud.githubusercontent.com/assets/16402401/22069287/e441fdec-dd65-11e6-8858-a95b20524233.png)
Screenshot shows processes, you can see on the taskbar that Streamlink is not currently running

**Comment by bastimeyer** [comment]
Can you please start the application from the shell with the `--enable-logging --v=1` parameters and see the output or error log?

**Comment by adrianAnyansi** [comment]
Took a while to trigger the bug again- might be 2-3 hours without opening streams. I ran the command in cmd and I'm assuming that the AppData\chrome_debug.log is the output. It didn't log on exit, but I was able to log when it tried to restart with the frozen process in the background. I'll keep trying with this.
First one is output when I was trying to trigger the error
[chrome_debug.txt](https://github.com/streamlink/streamlink-twitch-gui/files/715424/chrome_debug.txt)
2nd the error occurred around 7 but didnt see to log anything, and then I re-ran the command soon after that and it restarted the app
[chrome_debug2.txt](https://github.com/streamlink/streamlink-twitch-gui/files/715425/chrome_debug2.txt)

**Comment by adrianAnyansi** [comment]
Triggered it again using the _--enable-logging=stderr_ command and got this upon exit in the cmd
```
[45388:49216:0118/215354:VERBOSE1:dispatcher.cc(591)] Num tracked contexts: 1
[45388:49216:0118/215354:VERBOSE1:script_context.cc(121)] Destroyed context for extension
    extension id: kjlcknihpadniagphehmpojkkdigjp
    effective extension id: kjlcknihpadniagphehmpojkkdigjp
[45388:49216:0118/215354:VERBOSE1:script_context.cc(121)] Destroyed context for extension
    extension id:
    effective extension id:
```

Then it sent this message while I was typing
```
[47060:49736:0118/215920:VERBOSE1:policy_loader_win.cc(372)] Reading policy from the registry is disabled.
```

**Comment by bastimeyer** [comment]
There's nothing unusual to see in the logs... I'm unable to reproduce this behavior in any of my Windows VMs and I don't think that the project's code is responsible for this:
https://github.com/streamlink/streamlink-twitch-gui/blob/v1.1.1/src/app/initializers/nwjs.js#L89-L102
https://github.com/streamlink/streamlink-twitch-gui/blob/v1.1.1/src/app/services/NwjsService.js#L46-L52

This seems to be either a NW.js issue or something else on your system.
You could try deleting the config/cache folder (located in `%LOCALAPPDATA%`), but I don't think that this will do anything.

**Comment by adrianAnyansi** [comment]
Understood. As long as its not the project's fault I'll close the issue. Thanks!

**Comment by bastimeyer** [comment]
@Robby-

> This bug is very random and not 100% reproducible, or at least I haven't figured out the sequence of events to force this to happen at will. It does happen frequently, however. This didn't used to happen with the 0.x versions, though I can't say for sure with what exact version this issue started popping up.

Could anyone of you please run a development build of the application? Instructions can be found here:
https://github.com/streamlink/streamlink-twitch-gui/blob/v1.1.1/CONTRIBUTING.md#developing-and-building

Even if this sounds like a NW.js related issue, I'd like to know if this can be debugged. I'm unable to reproduce this in any of my Windows VMs. The `0.x` releases were using the legacy NW.js versions, so there might be something wrong in the new ones.

**Comment by Robby-** [comment]
Uh, yeah, I want to run a development build, but I do not really like to install this extra software on my Windows machine, and I do not currently have a Windows VM to do this on either. Maybe consider having nightly builds with this development mode enabled? I see there is already an issue about nightly builds (#360).

**Comment by bastimeyer** [comment]
@adrianAnyansi Could you please try the [latest release](https://github.com/streamlink/streamlink-twitch-gui/releases/latest) and see if the issue still exists? The `1.2.1` release now comes with NW.js `0.21.3` and Chromium `57.x`.

**Comment by adrianAnyansi** [comment]
Yes it does still occur on the latest release. I attempted to run a dev build but not having much luck. I've never worked with this tech so I don't know how to enable proper logging. DevTools from the app nor Chrome's 8888 port nor Ember's Chrome extension (does not activate) show any logging information.

**Comment by bastimeyer** [comment]
Can you check the `%TEMP%\streamlink-twitch-gui\logs` folder for log files? The `1.2.1` release now [creates logfiles by default](https://github.com/streamlink/streamlink-twitch-gui/wiki/Launch-parameters). Maybe an exception gets logged that prevents the process from exiting.

If you're already running a dev build, you can try setting a breakpoint [here](https://github.com/streamlink/streamlink-twitch-gui/blob/v1.2.1/src/app/main.js#L6) or [here](https://github.com/streamlink/streamlink-twitch-gui/blob/v1.2.1/src/app/initializers/nwjs.js#L91). Besides that, there's nothing really interesting that could be observed.

> Ember's Chrome extension

This your web browser's extension, not NW.js', unless it was explicitly added in the app's manifest.

**Comment by Robby-** [comment]
I also tried the `1.2.1` release and for me the problem also still exists. I used `--loglevel debug` but nothing is written to the log when trying to exit/close the program.

**Comment by MrPing1000** [comment]
I am also experiencing this.  It's a bit annoying.  (Win 7 64bit)

**Comment by qwaszer** [comment]
Same problem here: win7-64, v1.2.1, VLC

**Comment by Robby-** [comment]
Since upgrading to Streamlink Twitch GUI version 1.3.1 this issue hasn't occurred for me anymore.
So, I think this issue is actually fixed now, with the upgrade of NW.js to version 0.21.6.

**Comment by Robby-** [comment]
This issue has been randomly reoccurring again with version 1.4.1.

**Comment by Robby-** [comment]
@bastimeyer: Can you please re-open this issue, as it is reoccurring again with version `1.4.1`.

**Comment by bastimeyer** [comment]
In `1.3.0`, there was the following change, which switched from `nwWindow.close(true)` to `nwApp.quit()`:
b463d9eeb466d4e91224f7d48f9697c392a3eb52
It seems that this change had fixed the issue (unintentionally). I wasn't aware of it, otherwise I would have referred to this issue in the commit message.

This change broke something else on Linux due to a different NW.js bug, which causes calls to `nwApp.quit()` to not trigger any listeners attached to the `exit` event of nodejs. This is the reason, why
6ea0f44d8d060eb780206849c2b2cd209535024a + 56a964ee95885a4840647d62de3268c7affba726
were added *after* `1.3.1`, so that the notification helper process on Linux could be cleaned up on exit, which was a necessary change.

However, the notifications on Linux got changed again *before* `1.4.0` got released and since they don't require a helper process anymore,
3bca054b7bbbe303326e80f7318a75fba0a7c2c7
was added to revert the changes in favor of more graceful API calls. Since I was not aware of `nwApp.quit()` fixing this very issue here, I switched back to the old `nwWindow.close(true)` calls, because `nwApp.quit()` had caused the other issues related to the exit event listeners.

Before I'm going to change it again, I'd kindly ask you to test it first. I'll post a patch later when I get the time.

**Comment by Robby-** [comment]
I'd be happy to test it if you could please also build Windows 64-bit binaries with that patch applied for me to test with.

**Comment by bastimeyer** [comment]
This should be fixed now by b573b8d1bcb4260aa2c91e3e31e5e56ceae21498 (master branch)

> could please also build Windows 64-bit binaries with that patch applied for me

Sorry, but I don't upload any custom binaries here. You can build the app with just a few commands, nothing insanely difficult or time consuming, even on Windows. Just follow the guide described in README.md or [CONTRIBUTING.md](https://github.com/streamlink/streamlink-twitch-gui/blob/v1.4.1/CONTRIBUTING.md#developing-and-building).

Install git+NodeJS and run these lines
```bash
npm install -g yarn bower grunt-cli
git clone https://github.com/streamlink/streamlink-twitch-gui
cd streamlink-twitch-gui
yarn install
grunt release
```

Now that TravisCI has made cron-jobs available for everyone a few months ago, I could take a look at nightly builds again when I get the time.
